### DISCUSSION TITLE
Default version in the repository

### DISCUSSION THREAD

**Discussion by nbrucy** [ramses-organisation/ramses#53] [Polls]
For now the default version in the repository is the stable version. This make it easy for new user to download a (more) trusted version, but is somewhat mistake inducing to developers as Github currently defaults all pull request to the stable branch.

**Comment by nbrucy** [comment]
We started discussing this at the SNO meeting and here are some arguments for and against the change:

Cons:
- :-1: Make it harder for new user (`git clone --branch stable ` instead of `git clone`) - @blaizot 
- :-1: Maintainers can manually change PR that would go to `stable` to `dev` instead. - @nbrucy
- :-1: Because of the extra step to get it, ithe stable branch will become a "stale" branch that might not be used - @mtrebitsch 

Pros:
- :+1: Github default all PR toward the default branch - user can do mistakes - @nbrucy 
-  :+1:  It's the same for automatic tools: that cannot be changes? - @cphyc 
- :+1:  Stable version can be distributed in another way (tar on website / Release) - @duboisyo

(please tell me if I forgot something).

**Reply by mtrebitsch** [reply]
I think the last "pro" is actually a "con": if the stable version is not the default entry point, I'm pretty sure it will become a "stale" branch that might not be used (because of the extra step to get it). Happy to change my mind though!

**Comment by nbrucy** [comment]
Related question: do we want the default documentation to be the one of the stable of the dev version (the latter is more complete).
See issue #39 .

**Reply by ttepperg** [reply]
I'd say the default docs should correspond to the default branch. if the latter is the 'stable' branch, then ditto.

**Comment by nbrucy** [comment]
>[!WARNING]
> **Update**  It has been decided during the RUM that the `dev` branch should become the default branch, and the stable will be deleted.

`stable` will be replaced by regular releases with comprehensive changelog.
`dev` is now the default branch.

**Comment by ttepperg** [comment]
Hi Noé,

What does this mean in practice for my local copy of the repo?
What steps do I need to take to make sure a smooth transition?

Cheers

Thor

___________________________
Dr. Thor Tepper García
Astrophysicist
www.physics.usyd.edu.au/~tepper/

**Reply by nbrucy** [reply]
Hi Thor,

This will not affect your local copy. 

However, if you made some local development on top of the stable branch, you may want to keep them in sync with the dev branch (to profit from the various bugfixes and performance improvements that are continuously put in for instance).

There are several ways to do that. The best is to *rebase* your work on to the `dev` branch, but a classic merge could also do.
Don't hesitate to ask if you need more details.

Best,
Noé.

**Comment by ttepperg** [comment]
Hi Noé,

Thanks so much for your prompt and helpful reply.

> However, if you made some local development on top of the stable branch
That’s exactly the situation I’m in. I patched my version of the stable branch, but have not
’bothered’ to deal with the dev branch accordingly (until recently, I did not even know it existed TBH)

> you may want to keep them in sync with the dev branch (to profit from the various bugfixes and performance improvements that are continuously put in for instance).
> There are several ways to do that. The best is to rebase your work on to the dev branch, but a classic merge could also do

Not really sure how to do either without the risk of messing it all up.

Could you provide some more details / guidance?

So many thanks in advance!

Cheers

Thor

**Reply by nbrucy** [reply]
Sure. Again this is not mandatory, but it will help to keep up to date in the future.

Here is a way to do it:
1. First download a fresh copy of your repo: `git clone https://github.com/ttepperg/ramses.git`
2. Move into it `cd ramses`
3. Add the main repo as a new remote: `git remote add upstream https://github.com/ramses-organisation/ramses.git`
4. Fetch everything `git fetch --all`
5. Create a new dev branch: `git checkout -b dev`
6. Start rebasing your branch on top of the dev branch in upstream `git rebase upstream/dev`
7. You will encouter conflicts. No worries. Most of them are linked to modification you made to your gitignore, so maybe save the final version of it somewhere first and proceed with `git rebase --skip`. You can also skip the other conflicts (I checked). For completeness here is what to do if the conflict is more serious:
     1. Check the diff of the conflicting commit `git diff`
     2. Edit the conflicted files and select the version you want
     3. Once you solved all the commits, marked them as solved with `git add <file>`
     4. Proceed with `git continue` (you will be asked to validate the commit message)
8. Likewise, you can safely skip all empty commits
9. Your branch is now merged, you can publish it on github with `git push --set-upstream origin dev` and use it from now.

You old stable branch will remain untouched.

Cheers,
Noé.
